File: src/pattern/visitor/java/SizeVisitor.java
package pattern.visitor.java;

import java.util.Iterator;

public class SizeVisitor extends Visitor {
    private int totalSize = 0;

    @Override
    void visit(File file) {
        totalSize += file.getSize();
    }

    @Override
    void visit(Directory dir) {
        Iterator it = dir.iterator();
        while (it.hasNext()){
            Entry entry = (Entry)it.next();
            entry.accept(this);
        }
    }

    public int getSize(){
        return this.totalSize;
    }
}

File: src/pattern/visitor/kotlin/SizeVisitor.kt
package pattern.visitor.kotlin

class SizeVisitor: Visitor(){
    var totalSize:Int = 0
    override fun visit(file: File) {
        this.totalSize += file.getSize()
    }

    override fun visit(dir: Directory) {

        val it = dir.iterator()
        while (it.hasNext()){
            val entry = it.next()
            entry.accept(this)
        }
    }

    fun getSize(): Int{
        return this.totalSize
    }
}